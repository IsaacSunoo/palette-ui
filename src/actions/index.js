export const setError = error => ({
  type: "SET_ERROR",
  error
});

export const setLoading = loading => ({
  type: "SET_LOADING",
  loading
});

export const setPalette = palette => ({
  type: "SET_PALETTE",
  palette
});

export const setLock = colorID => ({
  type: "SET_LOCK",
  colorID
});

export const setModal = status => ({
  type: "SET_MODAL",
  status
});

export const setProjects = projects => ({
  type: "SET_PROJECTS",
  projects
});
