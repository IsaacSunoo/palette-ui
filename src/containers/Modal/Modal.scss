.Modal {
  position: fixed;
  z-index: 1;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgb(0,0,0);
  background-color: rgba(0,0,0,0.4);
}

.Modal-Content {
  background-color: #34a0e2;
  margin: 15% auto;
  padding: 2em;
  width: 60%;
  box-shadow: 0px 2px 7px 1px #231f20d9;
  border-radius: 5px;
  color:#fff;
}
