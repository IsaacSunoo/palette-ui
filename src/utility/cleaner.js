export const cleanProjectsPalettes = (projects, palettes) => {
  const cleanedData = projects.map(project => {
    const { id, name } = project;
    let projectData =  { id, name, palettes: [] }
    palettes.forEach(palette => {
      if(palette.project_id === project.id){
        projectData.palettes.push(palette)
      }
    });
    return projectData;
  });
  return cleanedData;
}
