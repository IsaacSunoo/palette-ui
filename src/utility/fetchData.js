export const fetchData = async (url, body) => {
  try {
    const response = await fetch(url, body)
    const data = await response.json()
    if(!response.ok) { throw new Error(`'Fetch Call Cannot Be Made' ${data}`)}
    return data;
  } catch (error) {
    return error;
  }
}
