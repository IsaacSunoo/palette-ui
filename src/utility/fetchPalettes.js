export const fetchPalettes = async (url) => {
  try {
    const response = await fetch(url)
    if(!response.ok) { throw new Error('Fetch Call Cannot Be Made')}
    const data = await response.json()
    return data;
  } catch (error) {
    console.log(error);
    // this.props.setError('Fetch Call Cannont')
  }
}
