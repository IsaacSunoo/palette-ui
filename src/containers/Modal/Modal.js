import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import * as actions from '../../actions/index';

class Modal extends Component {

  handleClose = () => {
    const { modalDisplay } = this.props;
    this.props.setModal(modalDisplay)
  };

  render() {


    return (
      <div className="Modal">
        <div className="Modal-Content">
          <section className="Modal-Header">
            <button onClick={this.handleClose}>
              Close
            </button>
          </section>
          <section className="Modal-Body">
            <h1>Save Project</h1>
            <p>Palette To Save</p>
          </section>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  currentPalette: state.currentPalette,
  modalDisplay: state.modalDisplay,
})

export const mapDispatchToProps = (dispatch) => ({
  setModal: (data) => dispatch(actions.setModal(data)),
})

Modal.propTypes = {
  currentPalette: PropTypes.array
};

export default connect(mapStateToProps, mapDispatchToProps)(Modal);
