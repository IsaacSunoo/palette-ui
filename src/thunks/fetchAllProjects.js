import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { cleanProjectsPalettes } from '../utility/cleaner';

export const fetchAllProjects = () => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const allProjects = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/projects');
      const allPalettes = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/palettes');
      dispatch(actions.setLoading(false));
      if(!allProjects.ok){
        throw Error(allProjects.statusText)
      }
      const combinedData = cleanProjectsPalettes(allProjects, allPalettes)
      dispatch(actions.setProjects(combinedData));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
