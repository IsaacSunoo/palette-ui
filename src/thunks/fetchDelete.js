import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { fetchOptions } from '../utility/fetchOptions';

export const fetchDelete = (url) => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const options = await fetchOptions('DELETE', {});
      await fetchData(url, options);
      dispatch(actions.setLoading(false));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
