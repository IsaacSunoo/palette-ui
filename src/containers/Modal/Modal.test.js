import React from 'react';
import { shallow } from 'enzyme'
import { Modal, mapStateToProps, mapDispatchToProps } from './Modal';

import { fetchOptions } from '../../utility/fetchOptions';
import { fetchData } from '../../utility/fetchData';
import { fetchAllProjects } from '../../thunks/fetchAllProjects';
jest.mock('../../thunks/fetchAllProjects');
jest.mock('../../utility/fetchData');
jest.mock('../../utility/fetchOptions');

import * as actions from '../../actions/index';

const mockAllProjects = [
  {
    id: 8,
    name: "Test Project",
    palettes: [
      {
        id: 15,
        name: "Mikes Pal",
        color_1: "#93125a",
        color_2: "#ffffff",
        color_3: "#f13470",
        color_4: "#80064e",
        color_5: "#423c52",
        project_id: 8,
      }
    ]
  }
]

const mockCurrentPalette = [
  { color: '#fffff', locked: false },
  { color: '#00000', locked: false },
  { color: '#231f20', locked: false },
  { color: '#e6e6e6', locked: false },
  { color: '#cccccc', locked: false }
]

describe('Modal', () => {

  describe('Modal Component', () => {
    let wrapper;
    let mockfetchAllProjects = jest.fn();
    let mockSetModal = jest.fn();
    let mockModalDisplay = true;

    beforeEach(() => {
      wrapper = shallow(<Modal allProjects={mockAllProjects}
                               currentPalette={mockCurrentPalette}
                               setModal={mockSetModal}
                               fetchAllProjects={mockfetchAllProjects}
                               modalDisplay={mockModalDisplay}
                        />
                      )
    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    });

    it('should have default state', () => {
      expect(wrapper.state()).toEqual({
        paletteName: '',
        projectId: ''
      });
    });

    it('should fetchData with corrects parameters', async () => {
      const mockUrl = 'http://localhost:3001/api/v1/palettes';
      const instance = wrapper.instance();
      const event = { preventDefault: () => { } };
      await instance.handleSendPalette(event);
      const mockOptions = await fetchOptions('POST', { name: '' });
      expect(fetchData).toHaveBeenCalledWith(mockUrl, mockOptions);
    });

    it('should change state when handleChange is changed', () => {
      const button = wrapper.find('.paletteInput');
      const mockEvent = { target: { name: 'paletteName', value: 'Project Title' } }
      button.simulate('change', mockEvent);
      expect(wrapper.state('paletteName')).toEqual('Project Title');
    });

    it('should call setModal when handleClose is invoked', () => {
      wrapper.instance().handleClose();
      expect(mockSetModal).toHaveBeenCalledWith(mockModalDisplay);
    });

  });

  describe('mapStateToProps', () => {
    it('should return an object with a modalDisplay value', () => {
      const mockState = {
        allProjects: mockAllProjects,
        currentPalette: mockCurrentPalette,
        modalDisplay: true,
        fakeState: 'Fake String',
      }
      const expected = {
        allProjects: mockAllProjects,
        currentPalette: mockCurrentPalette,
        modalDisplay: true,
      }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  });

  describe('mapDispatchToProps', () => {

    it('should call dispatch for fetchAllProjects', () => {
      //Setup
      const mockData = [
        { color: '#fffff', locked: false },
        { color: '#00000', locked: false },
        { color: '#231f20', locked: false },
        { color: '#e6e6e6', locked: false },
        { color: '#cccccc', locked: false }
      ];
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchAllProjects(mockData)
      //Execution
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchAllProjects(mockData);
      //Expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch for setPalette', () => {
      const mockData = false
      const mockDispatch = jest.fn();
      const actionToDispatch = actions.setModal(mockData)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setModal(mockData)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

  });

});
