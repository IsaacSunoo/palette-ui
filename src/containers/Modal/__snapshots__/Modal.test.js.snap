// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Modal Component should match the snapshot 1`] = `
<div
  className="Modal"
>
  <div
    className="Modal-Content"
  >
    <section
      className="Modal-Header"
    >
      <button
        onClick={[Function]}
      >
        <i
          className="fas fa-times"
        />
      </button>
    </section>
    <section
      className="Modal-Body"
    >
      <h3>
        Save Palette To Project
      </h3>
      <form
        onSubmit={[Function]}
      >
        <select
          name="projectId"
          onChange={[Function]}
          required={true}
          value=""
        >
          <option
            disabled={true}
            hidden={true}
            value=""
          >
            Choose Project
          </option>
          <option
            key="Test Project"
            name="projectId"
            value={8}
          >
            Test Project
          </option>
        </select>
        <input
          className="paletteInput"
          name="paletteName"
          onChange={[Function]}
          placeholder="Palette Name"
          required={true}
          type="text"
        />
        <button>
          Add Palette 
          <i
            className="fas fa-plus"
          />
        </button>
      </form>
      <section
        className="Modal-Palette"
      >
        <div
          className="Modal-Color"
          key="#fffff"
          style={
            Object {
              "backgroundColor": "#fffff",
            }
          }
        >
           
        </div>
        <div
          className="Modal-Color"
          key="#00000"
          style={
            Object {
              "backgroundColor": "#00000",
            }
          }
        >
           
        </div>
        <div
          className="Modal-Color"
          key="#231f20"
          style={
            Object {
              "backgroundColor": "#231f20",
            }
          }
        >
           
        </div>
        <div
          className="Modal-Color"
          key="#e6e6e6"
          style={
            Object {
              "backgroundColor": "#e6e6e6",
            }
          }
        >
           
        </div>
        <div
          className="Modal-Color"
          key="#cccccc"
          style={
            Object {
              "backgroundColor": "#cccccc",
            }
          }
        >
           
        </div>
      </section>
    </section>
  </div>
</div>
`;
